internal/internal_eager_activity.go

```diff
@@ -0,0 +1,157 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"fmt"
+	"sync"
+
+	"go.temporal.io/api/workflowservice/v1"
+)
+
+// eagerActivityExecutor is a worker-scoped executor for eager activities that
+// are returned from workflow task completion responses.
+type eagerActivityExecutor struct {
+	eagerActivityExecutorOptions
+
+	activityWorker *activityWorker
+	pendingCount   int // Only access under countLock
+	executingCount int // Only access under countLock
+	countLock      sync.Mutex
+}
+
+type eagerActivityExecutorOptions struct {
+	disabled  bool
+	taskQueue string
+	// If 0, there is no maximum
+	maxConcurrent int
+}
+
+// newEagerActivityExecutor creates a new worker-scoped executor without an
+// activityWorker set. The activityWorker must be set on the responding executor
+// before it will be able to execute activities.
+func newEagerActivityExecutor(options eagerActivityExecutorOptions) *eagerActivityExecutor {
+	return &eagerActivityExecutor{eagerActivityExecutorOptions: options}
+}
+
+func (e *eagerActivityExecutor) applyToRequest(
+	req *workflowservice.RespondWorkflowTaskCompletedRequest,
+) (amountReserved int) {
+	// Go over every command checking for activities that can be eagerly executed
+	for _, command := range req.Commands {
+		if attrs := command.GetScheduleActivityTaskCommandAttributes(); attrs != nil {
+			// If not present, disabled, or on a different task queue, we must mark as
+			// explicitly disabled
+			if e == nil || e.disabled || e.activityWorker == nil || e.taskQueue != attrs.TaskQueue.GetName() {
+				attrs.RequestEagerExecution = false
+			} else if attrs.RequestEagerExecution {
+				// If it has been requested, attempt to reserve one pending
+				attrs.RequestEagerExecution = e.reserveOnePendingSlot()
```

Comment:
nit: could count the number requested and reserve all at once to reduce the number of times you need to lock

Reply:
Yes, but then that'd require a bit more complication because it'd have to return how many it actually could reserve (possible it could only reserve half the activities for example). So basically it'd require multiple times through the command loop, but I may do that.